File: Week1_Design Patterns and Principles/Ex_1_Implementing_the_Singleton_Pattern/Code/AdvancedLogger.java


import java.time.LocalDateTime;
import java.time.format.DateTimeFormatter;

public class AdvancedLogger {

    private AdvancedLogger() {
        System.out.println("AdvancedLogger instance created at: " + getCurrentTimestamp());
    }

    private static class SingletonHelper {
        private static final AdvancedLogger INSTANCE = new AdvancedLogger();
    }

    public static AdvancedLogger getInstance() {
        return SingletonHelper.INSTANCE;
    }

    public void log(String message) {
        System.out.println("[" + getCurrentTimestamp() + "] ADVANCED LOG: " + message);
    }

    public void logError(String errorMessage) {
        System.err.println("[" + getCurrentTimestamp() + "] ADVANCED ERROR: " + errorMessage);
    }

    public void logWarning(String warningMessage) {
        System.out.println("[" + getCurrentTimestamp() + "] ADVANCED WARNING: " + warningMessage);
    }

    public void logInfo(String infoMessage) {
        System.out.println("[" + getCurrentTimestamp() + "] ADVANCED INFO: " + infoMessage);
    }

    private String getCurrentTimestamp() {
        return LocalDateTime.now().format(DateTimeFormatter.ofPattern("yyyy-MM-dd HH:mm:ss"));
    }

    public String getInstanceDetails() {
        return "AdvancedLogger instance: " + this.hashCode();
    }

    @Override
    protected Object clone() throws CloneNotSupportedException {
        throw new CloneNotSupportedException("Cannot clone singleton instance");
    }
}

File: Week1_Design Patterns and Principles/Ex_1_Implementing_the_Singleton_Pattern/Code/AdvancedLogger.cs
using System;

namespace SingletonDemo;


public sealed class AdvancedLogger : ICloneable
{
    private AdvancedLogger()
    {
        Console.WriteLine($"AdvancedLogger instance created at: {GetCurrentTimestamp()}");
    }


    private static class SingletonHolder
    {
        internal static readonly AdvancedLogger Instance = new();
    }

    public static AdvancedLogger Instance => SingletonHolder.Instance;

    public void Log(string message)        => Console.WriteLine($"[{GetCurrentTimestamp()}] ADVANCED LOG: {message}");
    public void LogError(string message)   => Console.Error.WriteLine($"[{GetCurrentTimestamp()}] ADVANCED ERROR: {message}");
    public void LogWarning(string message) => Console.WriteLine($"[{GetCurrentTimestamp()}] ADVANCED WARNING: {message}");
    public void LogInfo(string message)    => Console.WriteLine($"[{GetCurrentTimestamp()}] ADVANCED INFO: {message}");

    public string InstanceDetails => $"AdvancedLogger instance: {GetHashCode()}";
    private static string GetCurrentTimestamp() => DateTime.Now.ToString("yyyy-MM-dd HH:mm:ss");

   
    public object Clone() => throw new NotSupportedException("Cannot clone singleton instance");
}
